WIP: add KenLM support

dophist opened the pull request:
This PR adds kenlm support into Kaldi:

It is very similar to Kaldi's CARPA & OpenFST's G.fst(produced by arpa2fst), typical use cases are:

* decoding with super large LM (say biglm decoder or lookahead composition decoding)
* lattice rescoring with super large LM
* the KenLM FST wrapper can be used directly for LM online interpolation(or shallow fusion in nowadays fashion way)

The idea is simple: wrapping KenLM into Kaldi's _DeterministicOnDemandFst_ interface, then Kaldi users don't need to know if the underlying LM is KenLM or CARPA or OpenFST, treat it as a simple FST. When we are dealing with super large ngram models, KenLM outperforms the other two in terms of both speed & memory.

I wrote these code a couple of months ago, it was kind of a rush. And I've already tested the core functionality is working correctly in another codebase. But to get this into official Kaldi, there are still some work to do such as configure, makefile, cleaner unit test, and demo script etc.

@danpovey please have a quick look at this when you have time. If this is a desired enhancement, @naxingyu @chenguoguo and @jimbozhang said they can help with this feature.  At the moment, no need for detailed review, it's a merely a draft for now.

jtrmal commented:
Imo it would be good if you could add a binary for rescoring. Plus
modification of the makefiles and extras/install_kenlm.sh

On Fri, Aug 28, 2020 at 16:26 Jiayu DU <notifications@github.com> wrote:

> This PR adds kenlm support into Kaldi:
>
>
> It is very similar to Kaldi's CARPA & OpenFST's G.fst(produced by
> arpa2fst), typical use cases are:
>
>
>
>
>
>    - decoding with super large LM (say biglm decoder or lookahead
>    composition decoding)
>
>    - lattice rescoring with super large LM
>
>    - the KenLM FST wrapper can be used directly for LM online
>    interpolation(or shallow fusion in nowadays fashion way)
>
>
>
>
> The idea is simple: wrapping KenLM into Kaldi's *DeterministicOnDemandFst*
> interface, then Kaldi users don't need to know if the underlying LM is
> KenLM or CARPA or OpenFST, treat it as a simple FST. When we are dealing
> with super large ngram models, KenLM outperforms the other two in terms of
> both speed & memory.
>
>
> I wrote these code a couple of months ago, it was kind of a rush. And I've
> already tested the core functionality is working correctly in another
> codebase. But to get this into official Kaldi, there are still some work to
> do such as configure, makefile, cleaner unit test, and demo script etc.
>
>
> @danpovey <https://github.com/danpovey> please have a quick look at this
> when you have time. If this is a desired enhancement, @naxingyu
> <https://github.com/naxingyu> @chenguoguo <https://github.com/chenguoguo>
> and @jimbozhang <https://github.com/jimbozhang> said they can help with
> this feature. At the moment, no need for detailed review, it's a merely a
> draft for now.
>
>
>
>
> ------------------------------
>
>
>
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kaldi-asr/kaldi/pull/4247
>
>
>
>
> Commit Summary
>
>
>
>
>    - add KenLM support
>
>
>
>
>
>
> File Changes
>
>
>
>
>    -
>
>    *A*
>
>    src/lm/kenlm-fst.h
>    <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-81666e89258bbd0588eab68e60762d28>
>
>    (96)
>
>
>
>    -
>
>    *A*
>
>    src/lm/kenlm-test.cc
>    <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-4a9c85da9b49a0a70b153a5617b5de1b>
>
>    (41)
>
>
>
>    -
>
>    *A*
>
>    src/lm/kenlm.cc
>    <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-e59b82b620edfdeea4cec980acf66ecd>
>
>    (53)
>
>
>
>    -
>
>    *A*
>
>    src/lm/kenlm.h
>    <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-79d8dc599f58dc9e1df32a3c255c7b5e>
>
>    (62)
>
>
>
>
>
>
>
>
> Patch Links:
>
>
>
>
>    - https://github.com/kaldi-asr/kaldi/pull/4247.patch
>
>    - https://github.com/kaldi-asr/kaldi/pull/4247.diff
>
>
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUKYXY2GVRBFBQ2HHBQ6V3SC65APANCNFSM4QOF66TQ>
> .
>
>
>
>
>

dophist commented:
> Imo it would be good if you could add a binary for rescoring. Plus modification of the makefiles and extras/install_kenlm.sh
> […](#)
> On Fri, Aug 28, 2020 at 16:26 Jiayu DU ***@***.***> wrote: This PR adds kenlm support into Kaldi: It is very similar to Kaldi's CARPA & OpenFST's G.fst(produced by arpa2fst), typical use cases are: - decoding with super large LM (say biglm decoder or lookahead composition decoding) - lattice rescoring with super large LM - the KenLM FST wrapper can be used directly for LM online interpolation(or shallow fusion in nowadays fashion way) The idea is simple: wrapping KenLM into Kaldi's *DeterministicOnDemandFst* interface, then Kaldi users don't need to know if the underlying LM is KenLM or CARPA or OpenFST, treat it as a simple FST. When we are dealing with super large ngram models, KenLM outperforms the other two in terms of both speed & memory. I wrote these code a couple of months ago, it was kind of a rush. And I've already tested the core functionality is working correctly in another codebase. But to get this into official Kaldi, there are still some work to do such as configure, makefile, cleaner unit test, and demo script etc. @danpovey <https://github.com/danpovey> please have a quick look at this when you have time. If this is a desired enhancement, @naxingyu <https://github.com/naxingyu> @chenguoguo <https://github.com/chenguoguo> and @jimbozhang <https://github.com/jimbozhang> said they can help with this feature. At the moment, no need for detailed review, it's a merely a draft for now. ------------------------------ You can view, comment on, or merge this pull request online at: #4247 Commit Summary - add KenLM support File Changes - *A* src/lm/kenlm-fst.h <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-81666e89258bbd0588eab68e60762d28> (96) - *A* src/lm/kenlm-test.cc <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-4a9c85da9b49a0a70b153a5617b5de1b> (41) - *A* src/lm/kenlm.cc <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-e59b82b620edfdeea4cec980acf66ecd> (53) - *A* src/lm/kenlm.h <https://github.com/kaldi-asr/kaldi/pull/4247/files#diff-79d8dc599f58dc9e1df32a3c255c7b5e> (62) Patch Links: - https://github.com/kaldi-asr/kaldi/pull/4247.patch - https://github.com/kaldi-asr/kaldi/pull/4247.diff — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <#4247>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACUKYXY2GVRBFBQ2HHBQ6V3SC65APANCNFSM4QOF66TQ> .

Yes, that's definitely the plan.  And besides lattice rescoring, maybe another demo script on how to use kaldi to decode with super large language model(say more than 100G arpa) on the fly via lookahead composition.  Namely one for 2pass decoding, one for 1pass decoding.

naxingyu commented:
Todo
- [ ] decoder binary
- [ ] dependency checker & installer
- [ ] configuration
- [ ] decoding and rescoring script
- [ ] demo usage in a recipe (probably multi_cn)

Please comment for any supplement.

danpovey commented:
Sounds good!
Most important thing for me is, please try to anticipate any build problems
and provide workarounds, even if they disable KenLM support on some
platforms.
E.g. if KenLM won't work on ARM or with cross-compilation, we need to make
sure it doesn't break those compilation setups.


On Sat, Aug 29, 2020 at 11:30 AM Xingyu Na <notifications@github.com> wrote:

> Todo
>
>    - decoder binary
>    - dependency checker & installer
>    - configuration
>    - decoding and rescoring script
>    - demo usage in a recipe (probably multi_cn)
>
> Please comment for any supplement.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-683229635>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO552WBPSG3I2DJX5BLSDBY6LANCNFSM4QOF66TQ>
> .
>

chenguoguo commented:
Sounds good to me. I'll add more things to it.

dophist commented:
@chenguoguo @naxingyu , check the code change I just commited in src/latbin/lattice-lmrescore-const-arpa.cc , forget about the code correctness and leaking for the moment, it just shows an example how we can use KenLM in Kaldi.

danpovey commented on src/latbin/lattice-lmrescore-const-arpa.cc:
```diff
@@ -25,6 +25,8 @@
 #include "lm/const-arpa-lm.h"
 #include "util/common-utils.h"
 
+#include "lm/kenlm.h"
```
Cool.  Might be good idea, since  you're doing this in the same binary, to use an #include guard to optionally disable kenlm support, e.g. #ifndef NO_KENLM ... ?  The idea is that if we have trouble compiling KenLM on some platform, someone can add that to the CXXFLAGS in the Makefile.

dophist replied:
sure, we will pay extra attention on the install/configure/make system to isolate kenlm as a plugin module.

dophist commented:
sharing some info to get an idea how much kenlm can reduce the model file size(since kenlm uses on-disk data structure, model size basically means runtime memory footprint)

| arpa text file size | kenlm trie size without quantization | kenlm trie size with prob(8bits) & backoff(8bits) quantization|
| -- | --| -- |
| 119G | 32G |  18G |

In the compilation/building process for ngram this large, memory becomes a hard-limit & bottleneck when we use arpa2fst or arpa-to-const-arpa format. On the other hand, kenlm's compilation/training tool can operates in memory or on-disk, we can build large scale LM even on a PC with big enough hard disk.

jtrmal commented:
Yeah that's pretty cool. I have been having the same idea implemented for
srilm for a few years now and worked pretty well as well.

My doubts about these things were connected to the fact that I'm not sure
if huge arpas matter in the age of rnnlm. Any opinions on that?
Y.

On Sat, Aug 29, 2020 at 11:06 Jiayu DU <notifications@github.com> wrote:

>
>
> sharing some info to get an idea how much kenlm can reduce the model file
> size(since kenlm uses on-disk data structure, model size basically means
> runtime memory footprint)
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
> arpa text file size
> kenlm trie size without quantization
> kenlm trie size with prob(8bits) & backoff(8bits) quantization
>
>
>
>
> 119G
> 32G
> 18G
>
>
>
>
> In the compilation/building process for ngram this large, memory becomes a
> hard-limit & bottleneck when we use arpa2fst or arpa-to-const-arpa format.
> On the other hand, kenlm's compilation/training tool can operates in memory
> or on-disk, we can build large scale LM even on a PC with big enough hard
> disk.
>
>
>
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-683261812>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUKYXZPKLLVNFMPVTU5HZDSDDAIPANCNFSM4QOF66TQ>
> .
>
>
>

danpovey commented:
We should def. make use of it regardless.

On Sat, Aug 29, 2020 at 5:21 PM jtrmal <notifications@github.com> wrote:

> Yeah that's pretty cool. I have been having the same idea implemented for
> srilm for a few years now and worked pretty well as well.
>
> My doubts about these things were connected to the fact that I'm not sure
> if huge arpas matter in the age of rnnlm. Any opinions on that?
> Y.
>
> On Sat, Aug 29, 2020 at 11:06 Jiayu DU <notifications@github.com> wrote:
>
> >
> >
> > sharing some info to get an idea how much kenlm can reduce the model file
> > size(since kenlm uses on-disk data structure, model size basically means
> > runtime memory footprint)
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> >
> > arpa text file size
> > kenlm trie size without quantization
> > kenlm trie size with prob(8bits) & backoff(8bits) quantization
> >
> >
> >
> >
> > 119G
> > 32G
> > 18G
> >
> >
> >
> >
> > In the compilation/building process for ngram this large, memory becomes
> a
> > hard-limit & bottleneck when we use arpa2fst or arpa-to-const-arpa
> format.
> > On the other hand, kenlm's compilation/training tool can operates in
> memory
> > or on-disk, we can build large scale LM even on a PC with big enough hard
> > disk.
> >
> >
> >
> >
> > —
> > You are receiving this because you commented.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-683261812>,
> or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACUKYXZPKLLVNFMPVTU5HZDSDDAIPANCNFSM4QOF66TQ
> >
> > .
> >
> >
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-683263658>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4A3LCHMVPR2QK44RTSDDCCDANCNFSM4QOF66TQ>
> .
>

dophist commented:
@danpovey @jtrmal @chenguoguo @naxingyu there is a design issue worths some discussion here: should we allow vocabulary mismatch between base kaldi model and the rescoring kenlm model?

For example, during rescoring, its an ill-defined problem if kaldi yields a word to rescore lm that rescore lm doesn't recognize, report ERROR and quit is an option, rescore it with unk is another option.

In my past practice, it's always safe to enforce them to have strictly the same vocabulary and check this consistency as early as possible during runtime resource loading.  But human make errors, in the past several years, I've seen so many crashes due to mismatched vocab in rescoring(e.g. 1pass dynamic decoding with rescoring lm).

On the other hand, if we don't enforce vocab consistency, it brings great flexibility in production. For example, we don't need to rebuild all legacy resources all over again just because we want to add a new word to the vocabulary. Of course this flexibility is theoretically buggy, because ASR is defined as close-form vocabulary problem.

just a topic come to my mind when I was adding guard to vocab consistency check.

jtrmal commented:
IMO we should enforce the vocab is the same as a default.  Perhaps there
might be a flag that would do <unk> mapping as I assume it's not very
convenient to manipulate those huge lms to "fix them"). But the default
should be strict -- helps prevent confusion of users (on kaldi-help)
y.

On Sat, Aug 29, 2020 at 5:35 PM Jiayu DU <notifications@github.com> wrote:

> @danpovey <https://github.com/danpovey> @jtrmal
> <https://github.com/jtrmal> @chenguoguo <https://github.com/chenguoguo>
> @naxingyu <https://github.com/naxingyu> there is a design issue worths
> some discussion here: should we allow vocabulary mismatch between base
> kaldi model and the rescoring kenlm model?
>
> For example, during rescoring, its an ill-defined problem if kaldi yields
> a word to rescore lm that rescore lm doesn't recognize, report ERROR and
> quit is an option, rescore it with unk is another option.
>
> In my past practice, it's always safe to enforce them to have strictly the
> same vocabulary and check this consistency as early as possible during
> runtime resource loading. But human make errors, in the past several years,
> I've seen so many crashes due to mismatched vocab in rescoring(e.g. 1pass
> dynamic decoding with rescoring lm).
>
> On the other hand, if we don't enforce vocab consistency, it brings great
> flexibility in production. For example, we don't need to rebuild all legacy
> resources all over again just because we want to add a new word to the
> vocabulary. Of course this flexibility is theoretically buggy, because ASR
> is defined as close-form vocabulary problem.
>
> just a topic come to my mind when I was adding guard to vocab consistency
> check.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-683306083>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUKYXZMTRLGWWUCFCB63STSDEN43ANCNFSM4QOF66TQ>
> .
>

dophist commented:
ping...

danpovey commented:
This is a nice feature, but right now the problem is the installation procedure and upgrade path.
There is nothing in the tools/ data to install or download kenlm, and nothing in the configure script to deal with this.
If people just do 'git pull' after we merge this, they'll get a compilation error.   In fact the makefile hasn't even been modified
to add the new .o file corresponding to kenlm.cc.
We can't afford to inconvenience people that way.    Need to figure out a painless build and upgrade procedure.

dophist commented:
Thanks for the comment Dan, a bit busy these days and distracted from this work.  Since it's not very urgent, we will let you know when the PR is ready to merge.

danpovey commented:
Thanks!!

On Wed, Nov 11, 2020 at 9:50 PM Jiayu DU <notifications@github.com> wrote:

> Thanks for the comment Dan, a bit busy these days and distracted from this
> work. Since it's not very urgent, we will let you know when the PR is ready
> to merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-725434139>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLOYBBQWL62DK27NQUA3SPKJDXANCNFSM4QOF66TQ>
> .
>

danpovey commented:
By the way, making it optionally compiled would be nice as far as I am
concerned: I don't want to bulk up the codebase too much with obligatory
dependencies.

On Wed, Nov 11, 2020 at 9:58 PM Daniel Povey <dpovey@gmail.com> wrote:

> Thanks!!
>
> On Wed, Nov 11, 2020 at 9:50 PM Jiayu DU <notifications@github.com> wrote:
>
>> Thanks for the comment Dan, a bit busy these days and distracted from
>> this work. Since it's not very urgent, we will let you know when the PR is
>> ready to merge.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-725434139>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAZFLOYBBQWL62DK27NQUA3SPKJDXANCNFSM4QOF66TQ>
>> .
>>
>

jtrmal commented:
Could be added into make ext or separate target... Or both

On Wed, Nov 11, 2020 at 09:39 Daniel Povey <notifications@github.com> wrote:

> By the way, making it optionally compiled would be nice as far as I am
> concerned: I don't want to bulk up the codebase too much with obligatory
> dependencies.
>
> On Wed, Nov 11, 2020 at 9:58 PM Daniel Povey <dpovey@gmail.com> wrote:
>
> > Thanks!!
> >
> > On Wed, Nov 11, 2020 at 9:50 PM Jiayu DU <notifications@github.com>
> wrote:
> >
> >> Thanks for the comment Dan, a bit busy these days and distracted from
> >> this work. Since it's not very urgent, we will let you know when the PR
> is
> >> ready to merge.
> >>
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly, view it on GitHub
> >> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-725434139>,
> >> or unsubscribe
> >> <
> https://github.com/notifications/unsubscribe-auth/AAZFLOYBBQWL62DK27NQUA3SPKJDXANCNFSM4QOF66TQ
> >
> >> .
> >>
> >
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-725460146>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUKYX6IFYPLWVOK4XVYWCTSPKO3NANCNFSM4QOF66TQ>
> .
>

danpovey commented:
@kkm if you have time at some point to deal with the build issues around this, it would be great.. .up to you.

dophist commented:
MEMO & for convenience for reviewers:
added optional "--kenlm-root" option to __configure__(default off), this guarantees normal Kaldi building process unaffected by this PR.

When --kenlm-root is set (e.g ../tools/kenlm/), configure script propagates KENLM related variables into kaldi.mk, including:
1. KENLMROOT: 
* normally with value path/to/kaldi_root/tools/kenlm/, can be used by all sub-makefiles as a flag to check existence of kenlm module

2. KENLM_CXXFLAGS: 
* -DHAVE_KENLM: C++ level macro that we can use to check existence of kenlm module
* -I/path/to/kaldi_root/tools/kenlm : kenlm header including path
* -DKENLM_MAX_ORDER=6 : this is the only macro that is required by kenlm compilation, it's related to kenlm state struct static size.

3. KENLM_LDFLAGS:
* -L, -l, -rpath stuff

still need some works & checks, WIP

jtrmal commented:
I'd prefer configure option called 'enable-kenlm' that would, by default
set/propagated variables from tools/kenlm
kenlm-root might be suitable in case you wish to specify out-of-tree
build....
y.

On Thu, Nov 12, 2020 at 7:32 AM Jiayu DU <notifications@github.com> wrote:

> MEMO & for convenience for reviewers:
> added optional "--kenlm-root" option to *configure*(default off), this
> guarantees normal Kaldi building process unaffected by this PR.
>
> When --kenlm-root is set (e.g ../tools/kenlm/), configure script
> propagates KENLM related variables into kaldi.mk, including:
>
>    1. KENLMROOT:
>
>
>    - normally with value path/to/kaldi_root/tools/kenlm/, can be used by
>    all sub-makefiles as a flag to check existence of kenlm module
>
>
>    1. KENLM_CXXFLAGS:
>
>
>    - -DHAVE_KENLM: C++ level macro that we can use to check existence of
>    kenlm module
>    - -I/path/to/kaldi_root/tools/kenlm : kenlm header including path
>    - -DKENLM_MAX_ORDER=6 : this is the only macro that is required by
>    kenlm compilation, it's related to kenlm state struct static size.
>
>
>    1. KENLM_LDFLAGS:
>
>
>    - -L, -l, -rpath stuff
>
> still need some works & checks, WIP
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-726049647>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUKYX35CAR6HGWNY62BQ4TSPPIVRANCNFSM4QOF66TQ>
> .
>

dophist commented:
@jtrmal Good point, that makes perfect sense, will do.  Thanks!

dophist commented:
@danpovey @jtrmal Dan, Yenda, I think this PR is ready for review, to save you guys time, here is a short note of the code changes:
* kenlm is installed to *tools/kenlm/* via *tools/extras/install_kenlm_query_only.sh*
* kenlm is installed as a *runtime* library(query only mode) which means it can not be used to train an arpa.  This is because a full-build of kenlm (with ngram training support) depends on "eigen" and "boost", I believe it's impractical to integrate such heavy dependencies into Kaldi.
* kenlm is turned off by default unless users explicitly feed "--enable-kenlm" option to "configure"
    * once enabled, kenlm CXX and LD flags are accessible from kaldi.mk, Kaldi users can use these flags to compile & link with kenlm runtime library in src/xxxx/Makefile (refer to src/lm/Makefile)

at scripting level, two demo scripts are added:
* egs/wsj/s5/utils/train_arpa_with_kenlm.sh
demonstrates how to train an arpa from plain text corpus

* egs/wsj/s5/utils/build_kenlm_model_from_arpa.sh
demonstrate how to compile a standard arpa file into kenlm format

In terms of previously planned kenlm-based on-the-fly rescoring decoder & second-pass lattice rescoring bin, I'm afraid I need to find other time to work on, but not in this PR.  The kenlm-related foundation is already here, others may work out those bins and I'm happy to help.

dophist commented:
btw, the CodeFactor checking failure is because it stuck somewhere in my commit chain, not checking the newest code.

dophist commented:
It seems the rpath of KenLM is not properly propagated to binaries with dependency on libkaldi-lm, when kenlm is switched on.  Not quite sure, please have a look at this during review, I'll try to fix it this weekend.

danpovey commented:
Thanks a lot!
Sorry I am super busy, hope Yenda and Kiril will have time to look into
this.

On Fri, Nov 20, 2020 at 9:15 PM Jiayu DU <notifications@github.com> wrote:

> It seems the rpath of KenLM is not properly propagated to binaries with
> dependency on libkaldi-lm, when kenlm is switched on. Not quite sure,
> please have a look at this during review, I'll try to fix it this weekend.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4247#issuecomment-731163456>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4VD2QO2FQHPEZP22LSQZTVJANCNFSM4QOF66TQ>
> .
>

dophist commented:
Fixed.

Compared with Linux gcc, there is definitely something different related to rpath on MacOS.  I just removed kenlm support for MacOS for now. It is supposed to be used mostly in Linux anyway.

ready for reviews.

jtrmal commented on egs/wsj/s5/utils/train_arpa_with_kenlm.sh:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
+# 2020 author Jiayu DU
```
Appache 2.0 license please (all files).

jtrmal commented:
I had a brief glance over it -- it would be great if you could make even the shell scripts mostly respecting 80 characters per line (I didn't measure it but seems they are quite often much longer)

danpovey commented:
I think perhaps we could just merge as-is?

dophist commented:
Sorry for the delay. Yenda's review comments are addressed, I think we can merge this as-is, kenlm is by default switched off, shouldn't affect normal Kaldi build at all.  If there is any problem, please @ mention me and I'll respond quickly.

danpovey commented:
Thanks!!

nshmyrev commented:
@dophist sorry, this breaks cmake build since cmake doesn't define kenlm but still includes kenlm test into build. What is the best way around? Do we need to exclude kenlm from cmake or define required headers?

dophist commented:
> @dophist sorry, this breaks cmake build since cmake doesn't define kenlm but still includes kenlm test into build. What is the best way around? Do we need to exclude kenlm from cmake or define required headers?

Sorry for the inconvenience Nickolay, wasn't aware of that cmake is being used in Kaldi.  For quick workaround, excluding src/lm/kenlm.{h,cc}&kenlm-test.cc in CMake should be enough.  I'll start looking into Kaldi's CMake tomorrow, please help with the code review, I'll let you know when I open the fixing-PR.

dophist commented:
@nshmyrev the fixing-PR has been opened in (https://github.com/kaldi-asr/kaldi/pull/4448), if there is anything else, let me know : )

rezame commented:
@dophist 
Could you mind reporting the results (WER) of the model? 
How big text corpus? ppl? 
and resource memory and CPU time for decoding?  etc.
